### DISCUSSION TITLE
custom_config.h to (mostly) swap primary and secondary thumbs on Corne / ZMK?

### DISCUSSION THREAD

**Discussion by c-mart** [manna-harbour/miryoku#294] [Q&A] [answered]
On my Corne, the inner-most thumb keys feel like a more comfortable neutral position than the middle thumb keys, so I'd like to swap some (but not quite all) of their behavior.

| | left middle thumb | left inner thumb | right inner thumb | right middle thumb |
| --- | --- | --- | --- | --- |
| **default** | primary | secondary | secondary | primary |
| **desired** | secondary | primary | primary | secondary |

Tap behavior:

| | left middle thumb | left inner thumb | right inner thumb | right middle thumb |
| --- | --- | --- | --- | --- |
| **default** | space | tab | enter | backspace |
| **desired** | tab | space | backspace | enter |

Press-hold behavior:

| | left middle thumb | left inner thumb | right inner thumb | right middle thumb |
| --- | --- | --- | --- | --- |
| **default** | nav layer | mouse layer | symbol layer | number layer |
| **desired** | mouse layer | nav layer | number layer | symbol layer |

I also want to achieve:

- Unchanged behavior of left thumb keys while on number and symbol layers (to keep parens together, and 0 closer to the rest of the numpad).
- QWERTY as the default layout.

What I tried so far:

I made [this commit](https://github.com/c-mart/miryoku_zmk/commit/579fd2780a36fc5f9314b25429b1f16e43fdea8c#diff-2f6355536c15dc5fd88699e1d90ef1932388498cb1011943c16330ab52511cdeR17) from [these instructions](https://github.com/manna-harbour/miryoku/discussions/85). All I changed was swapping K33 with K34, and K35 with K36, in `MIRYOKU_LAYOUTMAPPING_CORNE`.

This got me partway there, except:

- Colemak became the default layout, even though I chose QWERTY in the workflow build options.
- It changed thumb key behavior on _all_ layers (I want to leave left thumb behvior unchanged on number and symbol layers).

No pressure to reply. I'll try to figure it out and leave the answer here for others.

**Comment by c-mart** [comment, marked as answer]
I did it: [custom_config.h](https://github.com/c-mart/miryoku_zmk/blob/0361569c6650d4e9f568ebea698cd209bddf478c/miryoku/custom_config.h)

Since I wanted to change behavior per-layer (not only re-map a hardware key across all layers), I needed to follow the "substitute layer" directions instead of the "change the mapping for all layers" directions. That means this solution should generalize to any keyboard, not just the Corne.

It made perfect sense after I got _all_ the layers in there:

![image](https://github.com/manna-harbour/miryoku/assets/13605043/adae622a-7bbb-4580-844b-985211c4ecd4)

The final line for each layer defines thumb key behavior. (I couldn't help myself from tweaking the inverted-T mapping while I was in there, but you can ignore that part.) I just had to swap the thumb bits around for the layers I wanted to change, e.g. `BSPC` and `RET`. Commit, push, build, flash, works great.

**Reply by manna-harbour** [reply]
Glad you found a solution!

> That means this solution should generalize to any keyboard, not just the Corne.

You can achieve the same result by either substituting a custom layer or by substituting a custom per-layer mapping.

With custom layers it will apply to all keyboards. That'll give you the same muscle memory on all keyboards, but also the same additional thumb movement on all keyboards. And other keyboards will have the middle thumb key closest to the thumb resting position, so this arrangement will be less comfortable.

With custom per-layer mapping it will be per-keyboard. That will give you an issue with muscle memory incompatibility between keyboards, but will restrict the changes to only where they are needed.

> It made perfect sense after I got all the layers in there:

Note that if you are not making any changes to a layer you don't need to define it.

**Comment by manna-harbour** [comment]
> On my Corne, the inner-most thumb keys feel like a more comfortable neutral position than the middle thumb keys, so I'd like to swap some (but not quite all) of their behavior.

See https://github.com/manna-harbour/miryoku/discussions/269#discussioncomment-8165513. The corne thumbs are a little cramped, but unless your resting thumb position is very far out you should be able to adapt to using the middle thumb key as home. The issue with changing it is that you'll lose muscle memory with Miryoku on other keyboards, and that you'll have to move more than 1u from the home thumb to the tertiary thumb key.

Also see https://github.com/manna-harbour/miryoku/discussions/159. The Base thumbs are also duplicated as non layer tap on other layers, so you'd need to either learn 2 positions, or re-arrange those layers too.

> It changed thumb key behavior on all layers (I want to leave left thumb behvior unchanged on number and symbol layers).

As you noted, changing the default mapping will affect all layers. You can also substitute custom mapping per layer. See the "To change the mapping per layer" section in the docs.

> Colemak became the default layout, even though I chose QWERTY in the workflow build options.

It should not affect the selection of alternative alpha arrangements.
